File: src/main/java/com/nik7/upgradecraft/UpgradeCraft.java
package com.nik7.upgradecraft;

import com.mojang.logging.LogUtils;
import com.nik7.upgradecraft.init.Config;
import com.nik7.upgradecraft.init.RegisterBlock;
import com.nik7.upgradecraft.init.RegisterBlockEntity;
import com.nik7.upgradecraft.init.RegisterItem;
import net.minecraftforge.eventbus.api.IEventBus;
import net.minecraftforge.fml.ModLoadingContext;
import net.minecraftforge.fml.common.Mod;
import net.minecraftforge.fml.config.ModConfig;
import net.minecraftforge.fml.javafmlmod.FMLJavaModLoadingContext;
import org.slf4j.Logger;

@Mod(UpgradeCraft.MODID)
public class UpgradeCraft {
    public static final String MODID = "upgradecraft";
    public static final Logger LOGGER = LogUtils.getLogger();

    public UpgradeCraft() {
        ModLoadingContext.get().registerConfig(ModConfig.Type.SERVER, Config.SERVER_CONFIG);
        IEventBus modEventBus = FMLJavaModLoadingContext.get().getModEventBus();

        RegisterBlock.register(modEventBus);
        RegisterItem.register(modEventBus);
        RegisterBlockEntity.register(modEventBus);

        // MinecraftForge.EVENT_BUS.register(ClientOnlyEventHandler.class);
    }

}

File: src/main/kotlin/com/nik7/upgradecraft/UpgradeCraft.kt
package com.nik7.upgradecraft

import com.nik7.upgradecraft.init.ModBlockEntities
import com.nik7.upgradecraft.init.ModBlocks
import com.nik7.upgradecraft.init.ModItems
import net.minecraft.client.Minecraft
import net.minecraft.client.renderer.ItemBlockRenderTypes
import net.minecraft.client.renderer.RenderType
import net.minecraftforge.fml.common.Mod
import net.minecraftforge.fml.event.lifecycle.FMLClientSetupEvent
import net.minecraftforge.fml.event.lifecycle.FMLDedicatedServerSetupEvent
import org.apache.logging.log4j.Level
import org.apache.logging.log4j.LogManager
import org.apache.logging.log4j.Logger
import thedarkcolour.kotlinforforge.forge.MOD_BUS
import thedarkcolour.kotlinforforge.forge.runForDist

/**
 * Main mod class. Should be an `object` declaration annotated with `@Mod`.
 * The modid should be declared in this object and should match the modId entry
 * in mods.toml.
 *
 * An example for blocks is in the `blocks` package of this mod.
 */
@Mod(UpgradeCraft.ID)
object UpgradeCraft {
    const val ID = "upgradecraft"

    // the logger for our mod
    val LOGGER: Logger = LogManager.getLogger(ID)

    init {
        LOGGER.log(Level.INFO, "Hello world!")

        // Register the KDeferredRegister to the mod-specific event bus
        ModBlocks.REGISTRY.register(MOD_BUS)
        ModItems.REGISTRY.register(MOD_BUS)
        ModBlockEntities.REGISTRY.register(MOD_BUS)

        val obj = runForDist(
            clientTarget = {
                MOD_BUS.addListener(UpgradeCraft::onClientSetup)
                Minecraft.getInstance()
            },
            serverTarget = {
                MOD_BUS.addListener(UpgradeCraft::onServerSetup)
                "test"
            })

        LOGGER.info(obj)
    }

    private fun onClientSetup(event: FMLClientSetupEvent) {
        event.enqueueWork {
            ItemBlockRenderTypes.setRenderLayer(ModBlocks.FLUID_FURNACE_BLOCK, RenderType.cutoutMipped())
        }
    }

    /**
     * Fired on the global Forge bus.
     */
    private fun onServerSetup(event: FMLDedicatedServerSetupEvent) {
        LOGGER.log(Level.INFO, "Server starting...")
    }
}
